tensorboard/components/tf_categorization_utils/categorizationUtils.ts

```diff
@@ -110,25 +122,102 @@ export function categorizeTags(
     query?: string): TagCategory[] {
   const tags = tf_backend.getTags(runToTag);
   const categories = categorize(tags, query);
-  const tagToRuns: {[tag: string]: string[]} = {};
-  tags.forEach(tag => {
-    tagToRuns[tag] = [];
-  });
-  selectedRuns.forEach(run => {
-    (runToTag[run] || []).forEach(tag => {
-      tagToRuns[tag].push(run);
-    });
-  });
+  const tagToRuns = createTagToRuns(_.pick(runToTag, selectedRuns));
+
   return categories.map(({name, metadata, items}) => ({
     name,
     metadata,
     items: items.map(tag => ({
       tag,
-      runs: tagToRuns[tag].slice(),
+      runs: tagToRuns.get(tag).slice(),
     })),
   }));
 }
 
+/**
+ * Creates grouping of the data based on selection from tf-data-selector. It
+ * groups data by prefixes of tag names and by tag names. Each group contains
+ * series, a tuple of experiment name and run name.
+ */
+export function categorizeSelection(
+    selection: tf_data_selector.Selection[], pluginName: string):
+    SeriesCategory[] {
+  const tagToSeries = new Map();
+  const searchTags = new Set();
+
+  selection.forEach(({experiment, runs, tagRegex}) => {
+    const runNames = runs.map(({name}) => name);
+    const selectedRunToTag = createRunToTagForPlugin(runs, pluginName);
+    const tagToSelectedRuns = createTagToRuns(selectedRunToTag);
+    const tags = tf_backend.getTags(selectedRunToTag);
+
+    const searchCategory = categorizeBySearchQuery(tags, tagRegex);
+    // list of matching tags.
+    searchCategory.items.forEach(tag => searchTags.add(tag));
+
+    // list of all tags that has selected runs.
+    tags.forEach(tag => {
+      const series = tagToSeries.get(tag) || [];
+      series.push(...tagToSelectedRuns.get(tag)
+          .map(run => ({experiment: experiment.name, run})));
+      tagToSeries.set(tag, series);
+    });
+  });
+
+  const searchCategory = {
+    name: selection.length == 1 ? selection[0].tagRegex : 'multi',
+    metadata: {
+      type: CategoryType.SEARCH_RESULTS,
+      validRegex: false,
+      universalRegex: false,
```

Comment:
Do the false values for these properties (validRegex, universalRegex) actually matter here?  Or will they get overridden and/or ignored later on somehow?